docs: add ESR 2.29 to 2.34 upgrade guide (#25910)
coder/coder: docs/install/releases/esr-2.29-2.34-upgrade.md

## Guide Overview

Coder provides Extended Support Releases (ESR) biannually. This guide walks
through upgrading from Coder 2.29 ESR to the planned Coder 2.34 ESR. It
summarizes key changes, highlights breaking updates, and provides a recommended
upgrade process.

Read more about the ESR release process
[here](./index.md#extended-support-release), and how Coder supports it.

## What's New in Coder 2.34

### Coder Agents

[Coder Agents](../../ai-coder/agents/index.md) is the long-term replacement for
Coder Tasks. Agents run the agent loop in the Coder control plane, store
conversation state in Coder, and can provision or attach to workspaces as needed.
This gives administrators stronger control over model access, credentials,
workspace selection, and auditability.

Between 2.29 and 2.34, Coder Agents gained multi-turn chat, file and image
attachments, workspace-aware prompts, MCP server integration, provider-native web
search, plan mode, explore subagents, chat labels, pinned chats, unread state,
chat debugging, and workspace terminal integration. Administrators also gained
platform controls for template allowlists, spend limits, model settings, MCP

> servers, Git providers, chat retention, and automatic chat archiving.

Reviewer: ```suggestion
Between 2.29 and 2.34, Coder Agents gained the following functionality:

- Multi-turn chat
- File and image attachments
- Workspace-aware prompts
- MCP server integration
- Provider-native web search
- Plan mode
- Subagent exploration
- Chat labels
- Pinned chats
- Unread state indicators
- Chat debugging
- Workspace terminal integration.

Administrators also now have platform controls for the following functionality:

- Template allowlists
- Spend limits
- Model settings
- MCP servers
- Git providers
- Chat retention
- Automatic chat archiving
```

I think listing them out is better. I avoid using prose lists for lists with more than 5 items.

Coder Tasks remains available during the support window, but it is being
replaced by Coder Agents. See the
[Tasks to Chats migration guide](../../ai-coder/agents/tasks-to-chats-migration.md)

> for API migration details.

Reviewer: If Coder Tasks is being replaced by Coder Agents, we should put a deprecation notice here, no?

### AI Gateway and AI Governance

AI Gateway, previously AI Bridge, matured into a broader governance and
observability layer for AI usage. It now supports proxy mode, OpenAI Responses
API interception, expanded Copilot support, ChatGPT support, custom Bedrock
endpoints, structured logs, client/session views, model filtering, multiple
providers of the same type, BYOK, and provider key policies.

AI Governance adds administrative controls around AI usage, including license and
seat visibility, AI session auditing, usage limits, group AI budgets, spend views,
model price tracking, and user or model-level controls. These features help
administrators understand who is using AI tools, which providers are being used,

> and how spend changes over time.

Reviewer: Convert these lists to bulleted lists.

### Agent Firewall

Agent Firewall, previously Agent Boundaries, moved from an early capability into
a stronger governance primitive for AI agents. It can audit and restrict network
access from agent processes, forward machine-readable logs to the control plane,
track usage, and use landjail mode for environments where changing Linux
capabilities is not practical.

> For more information, see the

Reviewer: ```suggestion
For more information, visit the
```

[Agent Firewall documentation](../../ai-coder/agent-firewall/index.md).

### Service Accounts

[Service accounts](../../admin/users/headless-auth.md) are now first-class users
for machine and automation use cases. They can be created through the API, CLI,
and dashboard, and they are designed to replace regular users with
`--login-type none` for CI/CD and other headless workflows.

Service accounts are a Premium feature and now integrate with workspace sharing,

> user and workspace filtering, organization membership, and role assignment.

Reviewer: ```suggestion
Service accounts are a [Premium](../../admin/licensing.md) feature 
and now integrate with workspace sharing, user and workspace filtering, 
organization membership, and role assignment.
```

### Templates, Prebuilds, and User Secrets

Template and workspace operations received several improvements:

- Terraform modules are cached per template version to reduce repeated downloads
  and make workspace starts more deterministic
- Prebuild claiming is more durable and idempotent
- Prebuild presets are validated with dynamic parameter validation
- `coder_env` supports `merge_strategy`
- User secrets can be created, encrypted, audited, and injected into workspaces

> - The dashboard warns about active prebuilds when duplicating templates

Reviewer: ```suggestion
- Terraform modules are cached per template version to reduce repeated downloads
  and make workspace starts more deterministic.
- Prebuild claiming is more durable and idempotent.
- Prebuild presets are validated with dynamic parameter validation.
- `coder_env` supports `merge_strategy`.
- User secrets can be created, encrypted, audited, and injected into workspaces.
- The dashboard warns about active prebuilds when duplicating templates.
```

Punctuation after bullets that complete a sentence or that are complete sentences on their own.

These changes reduce operational surprises for template authors, but templates
that assumed a clean Terraform module download on every build should be tested.
Templates that used the removed `coder_secret` Terraform data source must migrate
to user secrets through the Coder API, CLI, or workspace injection flow.

### Security and Networking

Coder added several security and networking controls between 2.29 and 2.34:

- OAuth2 external auth providers now support PKCE, and unknown providers default
  to PKCE unless explicitly disabled
- Secure auth cookies are now enabled automatically when `CODER_ACCESS_URL` uses
  HTTPS
- AI Gateway Proxy blocks CONNECT tunnels to private or reserved IP ranges, while
  always exempting the Coder access URL
- Workspace agents can disable reverse and local port forwarding through agent
  flags
- Authenticated request rate limiting is keyed by user instead of IP
- Kubernetes Gateway API `HTTPRoute` is supported as an alternative to Ingress
- Helm chart probes are more configurable, and Prometheus and pprof addresses can
  be overridden through chart environment values
- DERP TLS configuration is wired through the CLI, SDK, tailnet, VPN, agent, and

>   health checks

Reviewer: ```suggestion
- OAuth2 external auth providers now support PKCE, and unknown providers default
  to PKCE unless explicitly disabled.
- Secure auth cookies are now enabled automatically when `CODER_ACCESS_URL` uses
  HTTPS.
- AI Gateway Proxy blocks CONNECT tunnels to private or reserved IP ranges, while
  always exempting the Coder access URL.
- Workspace agents can disable reverse and local port forwarding through agent
  flags.
- Authenticated request rate limiting is keyed by user instead of IP address.
- Kubernetes Gateway API `HTTPRoute` is supported as an alternative to Ingress.
- Helm chart probes are more configurable, and Prometheus and pprof addresses can
  be overridden through chart environment values.
- DERP TLS configuration is wired through the CLI, SDK, tailnet, VPN, agent, and
  health checks.
```

### Operations and Scale

> Large deployments should see improvements in database, logging, and observability

Reviewer: ```suggestion
Large deployments should now have improvements in database, logging, and observability
```

configuration for audit logs, connection logs, API keys, and workspace agent
logs, dbpurge metrics, support bundle improvements, chatd metrics, agent first
connection duration metrics, and a `coder_build_info` metric. Coder also removed
several deprecated Prometheus metrics, so dashboards and alerts should be
reviewed before the upgrade.

Several expensive queries and write paths were optimized, including AI Gateway
session listing, audit and connection log counts, connection log batching,
provisioner job queue lookups, chat streaming, and coordinator peer mapping.

### CLI and Dashboard Enhancements

The CLI and dashboard gained smaller but meaningful workflow improvements:

- `coder create --no-wait` creates a workspace without waiting for startup
- `coder logs` provides easier access to logs
- `coder login token` prints the current session token for scripts and automation
- `coder support bundle` can infer the workspace from the environment
- `coder groups list -o json` now returns a flat JSON structure
- The dashboard includes user editing, service account management, group member
  filtering, role selection during user creation, improved accessibility, and

>   clearer confirmation flows for destructive or risky actions

Reviewer: ```suggestion
- `coder create --no-wait` creates a workspace without waiting for startup.
- `coder logs` provides easier access to logs.
- `coder login token` prints the current session token for scripts and automation.
- `coder support bundle` can infer the workspace from the environment.
- `coder groups list -o json` now returns a flat JSON structure.
- The dashboard includes user editing, service account management, group member
  filtering, role selection during user creation, improved accessibility, and
  clearer confirmation flows for destructive or risky actions.
```

| External provisioner daemons use the 2.29 provisionerd protocol                               | The provisionerd protocol changed for provisioner operations and file upload/download                                | Update external provisioner daemons to the matching 2.34 protocol. The protocol reserves removed fields such as `stop_modules`, `exp_reuse_terraform_workspace`, and `user_secrets`, and adds `DownloadFile`.                                          |
| Workspace sharing and chat sharing are only controlled through roles and feature availability | Deployment-wide sharing disable flags are available                                                                  | Review `CODER_DISABLE_WORKSPACE_SHARING` and `CODER_DISABLE_CHAT_SHARING` if your deployment needs to disable ACL-based sharing globally.                                                                                                              |

> | Helm chart health probes and observability bind addresses use older chart defaults            | Readiness and liveness probes have `enabled` toggles and more fields, and Prometheus/pprof addresses are overridable | Review custom Helm values for probe behavior and observability bindings. Prefer restricting pprof to a local address when exposing diagnostics.                                                                                                        |

Reviewer: Nit: the values in the first two columns should also have punctuation at the end of them.

## Upgrading

> The following are recommendations by the Coder team when performing the upgrade:

Reviewer: ```suggestion
The Coder team recommends taking the following steps when performing the upgrade:
```

Less passive and more direct.

| Helm chart health probes and observability bind addresses use older chart defaults.            | Readiness and liveness probes have `enabled` toggles and more fields, and Prometheus/pprof addresses are overridable. | Review custom Helm values for probe behavior and observability bindings. Prefer restricting pprof to a local address when exposing diagnostics.                                                                                                        |

## Upgrading

> [!NOTE]
> You can upgrade directly from 2.29 to 2.34. Stepping through intermediate
> minor versions is not required.
>
> This upgrade applies 108 database migrations. Coder applies them in order
> on startup. Most migrations are schema changes that complete quickly. Expect
> anywhere from under a minute to several minutes total, scaling with audit,
> connection, and chat log volume.
>
> Take a database backup before upgrading and validate the upgrade in a
> staging environment that mirrors production data volume.

> The Coder team recommends taking the following steps when performing the upgrade:

Reviewer 2: I don't think we have time to address this, but the phrasing here is really wonky for ESR users. We're saying to ESR customers, people on the long term support release, to upgrade and change all their stuff for a release they'll pop onto in 6 months or a year. I would figure this is more an update of changes they'll need to make before they get onto the _next_ ESR, as opposed to now

Coder Agents includes the following high-level functionality:

- Supports all major LLM providers
- Multi-turn chat
- Automatic workspace provisioning
- MCP server integration, personal skills, and administrator-managed skills
- ACL-based chat sharing across users and groups
- Admin-configurable advisor for planning and architecture guidance
- Plan and subagent explore modes
- Chat debugging
- Virtual desktop

Administrators have the following levers to configure appropriate access to various parts of Coder Agents:

- Template allow lists for agents
- BYOK for users
- Cost controls
- Configurable chat retention
- Automatic chat archiving
- Configurable system instructions
- Observability via AI Gateway, part of Coder's AI Governance Add-On

> [!CAUTION]
> Coder Tasks is officially deprecated in 2.34. It remains supported through the 2.34 ESR window

> > (into 2026) but receives no new features. Coder recommends migrating to Coder Agents

Author: ```suggestion
> Coder Tasks is officially deprecated in 2.34. It remains supported through the 2.34 ESR support window
> but receives no new features. Coder recommends migrating to Coder Agents
```

Coder Agents includes the following high-level functionality:

- Supports all major LLM providers
- Multi-turn chat
- Automatic workspace provisioning
- MCP server integration, personal skills, and administrator-managed skills
- ACL-based chat sharing across users and groups
- Admin-configurable advisor for planning and architecture guidance
- Plan and subagent explore modes
- Chat debugging
- Virtual desktop

Administrators have the following levers to configure appropriate access to various parts of Coder Agents:

- Template allow lists for agents
- BYOK for users
- Cost controls
- Configurable chat retention
- Automatic chat archiving
- Configurable system instructions
- Observability via AI Gateway, part of Coder's AI Governance Add-On

> [!CAUTION]
> Coder Tasks is officially deprecated in 2.34. It remains supported through the 2.34 ESR window

> > (into 2026) but receives no new features. Coder recommends migrating to Coder Agents

Reviewer 3: (into 2026) doesn't make much sense here to me

> and the Chats API now. See the [Tasks to Chats migration guide](../../ai-coder/agents/tasks-to-chats-migration.md)
> for API migration details.

### AI Gateway and AI Governance

AI Gateway, previously AI Bridge, matured into a broader governance and
observability layer for AI usage. It now supports:

> - [AI Gateway Proxy](../../ai-coder/ai-gateway/ai-gateway-proxy/index.md).

Reviewer 3: Nit: drop trailing periods since above list doesn't have them

Author: @nickvigilante for one last pass on whether we want to drop periods in all lists. Ideally, we should have a rule to lint these.

Reviewer: My philosophy on punctuation in lists is there should always be punctuation at the end of a sentence if:

- Each list item is a complete sentence.
- Each list item completes an otherwise incomplete sentence.

^^ This list itself is an example of where I'd expect to see punctuation.

The following is a list of my favorite foods:

- NY-style pizza
- Oxtail stew
- Pad Thai

^^ This list is fine not to have punctuation, because the preceding sentence is already complete on its own, and each item is a subject.

I won't block this for shipping this page, so feel free to merge as-is and I'll establish a comprehensive style guide and set of linters soon.

- OpenAI Responses API interception.
- Expanded Copilot and ChatGPT support.
- Custom Bedrock endpoints.
- Structured logs and client/session views.
- Model filtering.
- Multiple providers of the same type.
- [BYOK](../../ai-coder/ai-gateway/auth.md#bring-your-own-key-byok) and
  [key failover](../../ai-coder/ai-gateway/providers.md#key-failover).

[AI Governance](../../ai-coder/ai-governance.md) adds administrative controls
around AI usage:

- License and seat visibility.
- AI session auditing.
- Usage limits and group AI budgets.

> - Spend views and model price tracking.

Reviewer 3: These are not yet released.

```suggestion
```

Author: @david-fraley @jcjiang 
With the unification work, both of these do exist for Agnets, so removing from this section.

> - User or model-level controls.

Reviewer 3: No such feature exists at present.

```suggestion
```

  prebuild preset validation, `coder_env` merging, user secrets, and workspace
  starts with changed parameters.
- **Audit AI Gateway integrations:** Update experimental API routes, check
  permissions for interception/session data, migrate provider configuration
  from env vars to the database via `/ai/settings`, verify proxy mode behavior,

>   and review any injected MCP usage.

Author: @jcjiang @david-fraley can you confirm this?

Reviewer 2: Config does live in the ai/settings page ya. That's the unification stuff

Author: Agnet says that we now emit a warning if someone tries to set it via flags.

Author: So tried calling that out.
